[CI] QueryBuilderBWCIT testQueryBuilderBWC {cluster=UPGRADED} failing

This failure is related to #103285 but for release-tests. I think we need manage the feature flag a bit more for release tests.

**Build scan:**
https://gradle-enterprise.elastic.co/s/ktas4p2cwjbce/tests/:qa:full-cluster-restart:v8.12.0%23bwcTest/org.elasticsearch.upgrades.QueryBuilderBWCIT/testQueryBuilderBWC%20%7Bcluster=UPGRADED%7D


**Reproduction line:**
```
./gradlew ':qa:full-cluster-restart:v8.12.0#bwcTest' -Dtests.class="org.elasticsearch.upgrades.QueryBuilderBWCIT" -Dtests.method="testQueryBuilderBWC {cluster=UPGRADED}" -Dtests.seed=292D3F245DEA882B -Dtests.bwc=true -Dbuild.snapshot=false -Dtests.jvm.argline="-Dbuild.snapshot=false" -Dtests.locale=ms -Dtests.timezone=Japan -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
[Failure dashboard for `org.elasticsearch.upgrades.QueryBuilderBWCIT#testQueryBuilderBWC {cluster=UPGRADED}`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testQueryBuilderBWC%20%7Bcluster%3DUPGRADED%7D'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.upgrades.QueryBuilderBWCIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))


**Failure excerpt:**
```
java.lang.RuntimeException: An error occurred orchestrating test cluster.

  at __randomizedtesting.SeedInfo.seed([292D3F245DEA882B:2AB1D3E7B445EB3B]:0)
  at org.elasticsearch.test.cluster.local.DefaultLocalClusterHandle.execute(DefaultLocalClusterHandle.java:264)
  at org.elasticsearch.test.cluster.local.DefaultLocalClusterHandle.writeUnicastHostsFile(DefaultLocalClusterHandle.java:245)
  at org.elasticsearch.test.cluster.local.DefaultLocalClusterHandle.waitUntilReady(DefaultLocalClusterHandle.java:188)
  at org.elasticsearch.test.cluster.local.DefaultLocalClusterHandle.upgradeToVersion(DefaultLocalClusterHandle.java:161)
  at org.elasticsearch.test.cluster.local.DefaultLocalElasticsearchCluster.upgradeToVersion(DefaultLocalElasticsearchCluster.java:150)
  at org.elasticsearch.upgrades.ParameterizedFullClusterRestartTestCase.maybeUpgrade(ParameterizedFullClusterRestartTestCase.java:95)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:980)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.elasticsearch.test.cluster.local.DefaultLocalElasticsearchCluster$1.evaluate(DefaultLocalElasticsearchCluster.java:47)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

  Caused by: java.lang.RuntimeException: Elasticsearch process died while waiting for ports file. See console output for details.

    at org.elasticsearch.test.cluster.local.AbstractLocalClusterFactory$Node.lambda$waitUntilReady$0(AbstractLocalClusterFactory.java:274)
    at org.elasticsearch.test.cluster.util.Retry.lambda$retryUntilTrue$0(Retry.java:28)
    at java.lang.Thread.run(Thread.java:1583)

```

The actual root cause is (note this is already with the fix from #103341)
```
[2023-12-13T08:45:29,672][ERROR][o.e.b.Elasticsearch      ] [test-cluster-0] fatal exception while booting Elasticsearch
org.elasticsearch.ElasticsearchException: failed to load metadata
	at org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:120) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.node.Node.start(Node.java:311) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.bootstrap.Elasticsearch.start(Elasticsearch.java:458) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.bootstrap.Elasticsearch.initPhase3(Elasticsearch.java:251) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:73) ~[elasticsearch-8.13.0.jar:?]
Caused by: org.elasticsearch.gateway.CorruptStateException: org.elasticsearch.xcontent.XContentParseException: [-1:21839] [index_template] failed to parse field [index_template]
	at org.elasticsearch.gateway.PersistedClusterStateService.readXContent(PersistedClusterStateService.java:675) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.lambda$loadOnDiskState$6(PersistedClusterStateService.java:573) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.consumeFromType(PersistedClusterStateService.java:717) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadOnDiskState(PersistedClusterStateService.java:572) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:496) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:411) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createOnDiskPersistedState(GatewayMetaState.java:181) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createPersistedState(GatewayMetaState.java:149) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:107) ~[elasticsearch-8.13.0.jar:?]
	... 4 more
Caused by: org.elasticsearch.xcontent.XContentParseException: [-1:21839] [index_template] failed to parse field [index_template]
	at org.elasticsearch.xcontent.ObjectParser.throwFailedToParse(ObjectParser.java:616) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:611) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.fromXContent(ComposableIndexTemplateMetadata.java:68) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry$Entry.lambda$new$0(NamedXContentRegistry.java:54) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry.parseNamedObject(NamedXContentRegistry.java:147) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.support.AbstractXContentParser.namedObject(AbstractXContentParser.java:414) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.Metadata$Builder.fromXContent(Metadata.java:2686) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.readXContent(PersistedClusterStateService.java:673) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.lambda$loadOnDiskState$6(PersistedClusterStateService.java:573) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.consumeFromType(PersistedClusterStateService.java:717) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadOnDiskState(PersistedClusterStateService.java:572) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:496) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:411) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createOnDiskPersistedState(GatewayMetaState.java:181) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createPersistedState(GatewayMetaState.java:149) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:107) ~[elasticsearch-8.13.0.jar:?]
	... 4 more
Caused by: org.elasticsearch.xcontent.XContentParseException: [-1:21839] [index_template] failed to parse field [data_stream]
	at org.elasticsearch.xcontent.ObjectParser.throwFailedToParse(ObjectParser.java:616) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:611) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplate.parse(ComposableIndexTemplate.java:107) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.lambda$static$1(ComposableIndexTemplateMetadata.java:51) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.lambda$declareField$10(ObjectParser.java:431) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:609) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.fromXContent(ComposableIndexTemplateMetadata.java:68) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry$Entry.lambda$new$0(NamedXContentRegistry.java:54) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry.parseNamedObject(NamedXContentRegistry.java:147) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.support.AbstractXContentParser.namedObject(AbstractXContentParser.java:414) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.Metadata$Builder.fromXContent(Metadata.java:2686) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.readXContent(PersistedClusterStateService.java:673) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.lambda$loadOnDiskState$6(PersistedClusterStateService.java:573) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.consumeFromType(PersistedClusterStateService.java:717) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadOnDiskState(PersistedClusterStateService.java:572) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:496) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:411) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createOnDiskPersistedState(GatewayMetaState.java:181) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createPersistedState(GatewayMetaState.java:149) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:107) ~[elasticsearch-8.13.0.jar:?]
	... 4 more
Caused by: org.elasticsearch.xcontent.XContentParseException: [-1:21825] [data_stream_template] unknown field [failure_store]
	at org.elasticsearch.xcontent.ObjectParser.lambda$errorOnUnknown$2(ObjectParser.java:105) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:299) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.lambda$declareField$10(ObjectParser.java:431) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:609) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplate.parse(ComposableIndexTemplate.java:107) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.lambda$static$1(ComposableIndexTemplateMetadata.java:51) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.lambda$declareField$10(ObjectParser.java:431) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:609) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.fromXContent(ComposableIndexTemplateMetadata.java:68) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry$Entry.lambda$new$0(NamedXContentRegistry.java:54) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry.parseNamedObject(NamedXContentRegistry.java:147) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.support.AbstractXContentParser.namedObject(AbstractXContentParser.java:414) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.Metadata$Builder.fromXContent(Metadata.java:2686) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.readXContent(PersistedClusterStateService.java:673) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.lambda$loadOnDiskState$6(PersistedClusterStateService.java:573) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.consumeFromType(PersistedClusterStateService.java:717) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadOnDiskState(PersistedClusterStateService.java:572) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:496) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:411) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createOnDiskPersistedState(GatewayMetaState.java:181) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createPersistedState(GatewayMetaState.java:149) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:107) ~[elasticsearch-8.13.0.jar:?]
	... 4 more

```

Looking more carefully in the logs I see that this is not a problem of communication between nodes, but the upgrade on one node is failing because it's transitioning from a version with the feature flag enabled  (8.12.0-SNAPSHOT) to a version with the feature flag disabled (8.13.0). See the logs:

```
[2023-12-21T20:02:53,699][INFO ][o.e.n.Node               ] [test-cluster-1] version[8.12.0-SNAPSHOT], ....
.......
[2023-12-21T20:03:11,703][INFO ][o.e.n.Node               ] [test-cluster-1] stopping ...
[2023-12-21T20:03:11,705][INFO ][o.e.x.w.WatcherService   ] [test-cluster-1] stopping watch service, reason [shutdown initiated]
[2023-12-21T20:03:34,633][INFO ][o.e.n.Node               ] [test-cluster-1] version[8.13.0], ....
.....
[2023-12-21T20:03:49,730][ERROR][o.e.b.Elasticsearch      ] [test-cluster-1] fatal exception while booting Elasticsearch
org.elasticsearch.ElasticsearchException: failed to load metadata
	at org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:120) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.node.Node.start(Node.java:311) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.bootstrap.Elasticsearch.start(Elasticsearch.java:458) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.bootstrap.Elasticsearch.initPhase3(Elasticsearch.java:251) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:73) ~[elasticsearch-8.13.0.jar:?]
Caused by: org.elasticsearch.gateway.CorruptStateException: org.elasticsearch.xcontent.XContentParseException: [-1:14043] [index_template] failed to parse field [index_template]
	at org.elasticsearch.gateway.PersistedClusterStateService.readXContent(PersistedClusterStateService.java:675) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.lambda$loadOnDiskState$6(PersistedClusterStateService.java:574) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.consumeFromType(PersistedClusterStateService.java:717) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadOnDiskState(PersistedClusterStateService.java:573) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:497) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:412) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createOnDiskPersistedState(GatewayMetaState.java:181) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createPersistedState(GatewayMetaState.java:149) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:107) ~[elasticsearch-8.13.0.jar:?]
	... 4 more
Caused by: org.elasticsearch.xcontent.XContentParseException: [-1:14043] [index_template] failed to parse field [index_template]
	at org.elasticsearch.xcontent.ObjectParser.throwFailedToParse(ObjectParser.java:616) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:611) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.fromXContent(ComposableIndexTemplateMetadata.java:68) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry$Entry.lambda$new$0(NamedXContentRegistry.java:54) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry.parseNamedObject(NamedXContentRegistry.java:147) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.support.AbstractXContentParser.namedObject(AbstractXContentParser.java:414) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.Metadata$Builder.fromXContent(Metadata.java:2686) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.readXContent(PersistedClusterStateService.java:673) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.lambda$loadOnDiskState$6(PersistedClusterStateService.java:574) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.consumeFromType(PersistedClusterStateService.java:717) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadOnDiskState(PersistedClusterStateService.java:573) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:497) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:412) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createOnDiskPersistedState(GatewayMetaState.java:181) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createPersistedState(GatewayMetaState.java:149) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:107) ~[elasticsearch-8.13.0.jar:?]
	... 4 more
Caused by: org.elasticsearch.xcontent.XContentParseException: [-1:14043] [index_template] failed to parse field [data_stream]
	at org.elasticsearch.xcontent.ObjectParser.throwFailedToParse(ObjectParser.java:616) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:611) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplate.parse(ComposableIndexTemplate.java:107) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.lambda$static$1(ComposableIndexTemplateMetadata.java:51) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.lambda$declareField$10(ObjectParser.java:431) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:609) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.fromXContent(ComposableIndexTemplateMetadata.java:68) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry$Entry.lambda$new$0(NamedXContentRegistry.java:54) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry.parseNamedObject(NamedXContentRegistry.java:147) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.support.AbstractXContentParser.namedObject(AbstractXContentParser.java:414) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.Metadata$Builder.fromXContent(Metadata.java:2686) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.readXContent(PersistedClusterStateService.java:673) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.lambda$loadOnDiskState$6(PersistedClusterStateService.java:574) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.consumeFromType(PersistedClusterStateService.java:717) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadOnDiskState(PersistedClusterStateService.java:573) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:497) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:412) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createOnDiskPersistedState(GatewayMetaState.java:181) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createPersistedState(GatewayMetaState.java:149) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:107) ~[elasticsearch-8.13.0.jar:?]
	... 4 more
Caused by: org.elasticsearch.xcontent.XContentParseException: [-1:14029] [data_stream_template] unknown field [failure_store]
	at org.elasticsearch.xcontent.ObjectParser.lambda$errorOnUnknown$2(ObjectParser.java:105) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:299) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.lambda$declareField$10(ObjectParser.java:431) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:609) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplate.parse(ComposableIndexTemplate.java:107) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.lambda$static$1(ComposableIndexTemplateMetadata.java:51) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.lambda$declareField$10(ObjectParser.java:431) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:609) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:166) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.ComposableIndexTemplateMetadata.fromXContent(ComposableIndexTemplateMetadata.java:68) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry$Entry.lambda$new$0(NamedXContentRegistry.java:54) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.NamedXContentRegistry.parseNamedObject(NamedXContentRegistry.java:147) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.xcontent.support.AbstractXContentParser.namedObject(AbstractXContentParser.java:414) ~[elasticsearch-x-content-8.13.0.jar:?]
	at org.elasticsearch.cluster.metadata.Metadata$Builder.fromXContent(Metadata.java:2686) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.readXContent(PersistedClusterStateService.java:673) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.lambda$loadOnDiskState$6(PersistedClusterStateService.java:574) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.consumeFromType(PersistedClusterStateService.java:717) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadOnDiskState(PersistedClusterStateService.java:573) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:497) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.PersistedClusterStateService.loadBestOnDiskState(PersistedClusterStateService.java:412) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createOnDiskPersistedState(GatewayMetaState.java:181) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.createPersistedState(GatewayMetaState.java:149) ~[elasticsearch-8.13.0.jar:?]
	at org.elasticsearch.gateway.GatewayMetaState.start(GatewayMetaState.java:107) ~[elasticsearch-8.13.0.jar:?]
	... 4 more
[2023-12-21T20:03:49,738][INFO ][o.e.n.Node               ] [test-cluster-1] stopping ...
[2023-12-21T20:03:49,755][INFO ][o.e.n.Node               ] [test-cluster-1] stopped
[2023-12-21T20:03:49,755][INFO ][o.e.n.Node               ] [test-cluster-1] closing ...
[2023-12-21T20:03:49,762][INFO ][o.e.n.Node               ] [test-cluster-1] closed
[2023-12-21T20:03:49,764][INFO ][o.e.x.m.p.NativeController] [test-cluster-1] Native controller process has stopped - no new native processes can be started
```

This is not desirable and enabling the feature flag is not the solution here. We need to figure out how to make this compatible without exposing the `failure_store` option to the user.

--- a/qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/upgrades/QueryBuilderBWCIT.java
+++ b/qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/upgrades/QueryBuilderBWCIT.java
@@ -39,6 +39,7 @@ import org.elasticsearch.index.query.functionscore.FunctionScoreQueryBuilder;
 import org.elasticsearch.index.query.functionscore.RandomScoreFunctionBuilder;
 import org.elasticsearch.search.SearchModule;
 import org.elasticsearch.test.cluster.ElasticsearchCluster;
+import org.elasticsearch.test.cluster.FeatureFlag;
 import org.elasticsearch.test.cluster.local.LocalClusterConfigProvider;
 import org.elasticsearch.test.cluster.local.distribution.DistributionType;
 import org.elasticsearch.xcontent.XContentBuilder;
@@ -76,6 +77,7 @@ public class QueryBuilderBWCIT extends ParameterizedFullClusterRestartTestCase {
         .version(getOldClusterTestVersion())
         .nodes(2)
         .setting("xpack.security.enabled", "false")
+        .feature(FeatureFlag.FAILURE_STORE_ENABLED)
         .apply(() -> clusterConfig)
         .build();
 

